base/tuple.jl

```diff
@@ -352,3 +352,72 @@ any(x::Tuple{Bool, Bool, Bool}) = x[1]|x[2]|x[3]
 Returns an empty tuple, `()`.
 """
 empty(x::Tuple) = ()
+
+## Linked-list representation of a tuple. Inferrable even for Type elements.
+
+struct TupleLLEnd end
+struct TupleLL{T, Rest}
+    head::T    # car
+    rest::Rest # cdr
+    TupleLL(x, rest::TupleLL) where {} = new{Core.Typeof(x), typeof(rest)}(x, rest) # (cons x rest)
+    TupleLL(x, rest::TupleLLEnd) where {} = new{Core.Typeof(x), typeof(rest)}(x, rest) # (cons x nil)
+    TupleLL(x) where {} = new{Core.Typeof(x), TupleLLEnd}(x, TupleLLEnd()) # (list x)
+    TupleLL() where {} = new{TupleLLEnd, TupleLLEnd}(TupleLLEnd(), TupleLLEnd())
```

Comment:
This would be a Tuple of one element (of value `TupleLLEnd()`)

Reply:
Then there is no way to construct an empty TupleLL.

Reply:
yes, like a lisp (the empty list is the same nil / an empty TupleLL is nothing)

Reply:
You mean literally `nothing`? I.e., not actually `<:TupleLL`?

I could switch to that (seems more annoying, though). We'll have to use a different value than `nothing` for "not yet set" for the `indexing` field.